### DISCUSSION TITLE
Is it ok to use canvas as Texture?

### DISCUSSION THREAD

**Discussion by humbleCoder1990** [pixijs/pixijs#7655] [Q&A]
![image](https://user-images.githubusercontent.com/87828192/127735089-ab63b13b-49f2-47a3-9df3-fcf45abc44e3.png)

I want to make a helper function to make a texture from canvas like this code.
Because cavnas 2d looks better than Graphics like a rounded border.
And I found I can pass canvas to Texture.from(canvas)
But is it ok about a variety of aspects for Pixi app? 
like performance, or future release etc..
Also would calling document.createElement('canvas') be no memory leaks to make a texture?
I want to use a lot of 'document.createElement('canvas')' for drawing side in Pixi app. (runtime drawing also)
Please tell me your opinions.
Thanks!

Lee

**Comment by ivanpopelyshev** [comment]
Generally, yes, you can use it.

As for how performant is it - yes, I use it for same canvas , when i need AAA for my shapes. Alternative is https://github.com/pixijs/graphics-smooth but its not covering all the cases yet.

As for memory leaks... I dont know how to describe it. You have to know whats "texImage2d" and that it actually copies data from canvas2d entirely on gpu-side if its accelerated. I looked in sources of chromium, im sure in that method.

the problem with "from" method is that it remembers canvas link in cache, so if you dont destroy texture - canvas will exist in js side till the end of time :) If you destroy it when you actually need it , and pixi gc destroys glTexture - then you wont be use elements with this texture at all. Because of all internals i know about - i cannot guarantee anything about it if you dont know how it works.

so there are
1. canvas
2. cache entry linking baseTexture->CanvasResource->canvas.
3. baseTexture->glTexture , when itst uploaded, can be cleared by pixi texture gc if its not used for several minutes. Its a GL resource, its a videomem, cant be just removed by js gc!

Another downside is that you cannot use same canvas for multiple elements unless you draw them nearby and separate Texture (frame rectangles).

Yes, it should be a big article. I cannot make it now :(

**Reply by humbleCoder1990** [reply]
If I destory the texture, will the Sprite using this texture also be broken? (no display state, or something else invaild)

**Reply by eXponenta** [reply]
Sprite - mesh with simple material and texture.
How you think?
